--- a/dwm.c
+++ b/dwm.c
@@ -442,7 +442,7 @@
 	if (ev->window == selmon->barwin) {
 		i = x = 0;
 		do
-			x += TEXTW(tags[i]);
+			x += MAX(TEXTW(tags[i]), tagwidth);
 		while (ev->x >= x && ++i < LENGTH(tags));
 		if (i < LENGTH(tags)) {
 			click = ClkTagBar;
@@ -729,7 +729,7 @@
 	}
 	x = 0;
 	for (i = 0; i < LENGTH(tags); i++) {
-		w = TEXTW(tags[i]);
+		w = MAX(TEXTW(tags[i]), tagwidth);
 		drw_setscheme(drw, scheme[urg & 1 << i ? SchemeUrg
 			: m->tagset[m->seltags] & 1 << i ? SchemeSel
 			: occ & 1 << i ? SchemeOcc
